### DISCUSSION TITLE
How to make a HTTP request synchronously in a coroutine

### DISCUSSION THREAD

**Discussion by rsp4jack** [cc-tweaked/CC-Tweaked#1264] [Q&A] [answered]
I have tried this code:
```lua
require("coroutine")

local imdone = false

local function threadhttp()
    local request = http.get("https://example.tweaked.cc")
    print(request.readAll()) -- (2)
    request.close()
    imdone = true
end

local coro = coroutine.create(threadhttp)

while imdone == false do
    coroutine.resume(coro)
    print(coroutine.status(coro))
    sleep(0.1)
end
```
Output:
```
suspended
suspended
...(repeat)
(Ctrl+T)
```
HTTP server has already responded but code will never be executed to (2).
How to make a HTTP request synchronously in a coroutine?

CC:T Version: cc-tweaked-1.18.2-1.101.0

**Comment by SquidDev** [comment, marked as answer]
Aaaah, that's not quite how coroutines and the event system work. You need to resume the coroutine with events you've pulled from further up the coroutine tree. See [this recent discussion](https://github.com/cc-tweaked/CC-Tweaked/discussions/1253) for a example of how you should be resuming coroutines!

**Reply by SquidDev** [reply]
Though in this case, you could probably get away without using http.request` and the events directly instead.

**Reply by Lupus590** [reply]
Adding to this, you should avoid using sleep as it will eat events that are not its timer event. Use `os.pullEvent`, `os.pulleventRaw` or `coroutine.yield` instead.

**Reply by rsp4jack** [reply]
Now it works:
```lua
require("coroutine")
local pretty = require "cc.pretty"

local imdone = false

local function threadhttp()
    local request = http.get("https://example.tweaked.cc")
    print(request.readAll())
    request.close()
end

local coro = coroutine.create(threadhttp)

local ok, filter = coroutine.resume(coro)
while coroutine.status(coro) ~= "dead" do
pretty.pretty_print(filter)
  local event = table.pack(os.pullEvent(filter))
  pretty.pretty_print(event)
  ok, filter = coroutine.resume(coro, table.unpack(event, 1, event.n))
end
if not ok then error(filter, 0) end
```
Thanks for your explanation!

**Reply by rsp4jack** [reply]
BTW, it is confusing to a beginner. I think there should be some guides about coroutine and event system in docs.

**Reply by SquidDev** [reply]
[BB's guide to coroutines](http://www.computercraft.info/forums2/index.php?/topic/25670-bbs-guide-to-coroutines/) is probably the best introduction to the whole thing.

The thing is coroutines are pretty niche. 90% of the time you don't need to worry about coroutines, and when you do, almost all of the time you can just get away with using the `parallel` API, so there's definitely more important topics we should document first.
